doc/developer/design/20240625_source_versioning__table_from_sources.md

```diff
@@ -0,0 +1,537 @@
+and output to the correct `output_index`.
+
+#### Migration of source statements and collections
+
+We would generate a catalog migration to generate new `CREATE TABLE` statements where
+necessary and shuffle the collections being pointed at by each statement. The intent
+is to preserve the names associated with each collection even if the statement that
+identifies each collection is updated. The migration would do the following:
+
+-   For existing sources where the primary relation has the data (e.g. Kafka sources):
+    -   Generate a new `CREATE TABLE` statement that is tied to the current primary collection
+        and name it `<source>`
+    -   Change the `CREATE SOURCE` statement to point to the current progress collection and
+        change the source name to `<source>_progress`
+    -   Drop the existing `CREATE SUBSOURCE <source>_progress` statement since this will no
+        longer have a collection to own
+-   For existing multi-output sources (e.g. Postgres & MySQL sources):
+    -   Convert subsources to `CREATE TABLE` statements with the same name
+    -   Change the `CREATE SOURCE` statement to point to the current progress collection and
+        change the source name to `<source>_progress`
+    -   Drop the existing `CREATE SUBSOURCE <source>_progress` statement since this will no
+        longer have a collection to own
+    -   Drop the existing top-level collection tied to the source, since this is already
+        unused and will no longer be owned by any statement
+
+#### Exposing available upstream references to users
+
+Since creating a source will no longer automatically create any subsources, users that
+do not already know all the upstream references available in their system may find it difficult
+to know which `CREATE TABLE .. FROM SOURCE` statements they could use to ingest their
+upstream data. Additionally, the Web Console will need a way to present a dynamic UI
+to users after they have created a source to allow them to create any downstream tables
+from a selection of the available upstream references of that source.
+
+We have two options available to us for exposing this information to users:
+
+1.  Add a new statement to fetch from the upstream all possible 'upstream references' available
+    for ingestion.
+2.  Create a new system table that periodically updates with the set of available upstream
+    references in each sources, and a new statement to trigger a refresh of this system table.
+
+The 2nd option is more complex to implement, but presents the information in a form that
+is more similar to other similar information in Materialize. However since the data can
+become out of date, it's important that the refresh statement is highlighted prominently
+in user-facing documentation and related guides.
+
+The proposal for the 2nd option (a new system table):
+
+Introduce a `mz_catalog.mz_available_source_references` with the following columns:
```

Comment:
Not at all married to this, but in `mz_source_statistics` we use "known (as in `offset_known`) to represent the idea of things that are known to Materialize but may not be perfectly up to date.

```suggestion
Introduce a `mz_catalog.mz_known_source_references` with the following columns:
```

Reply:
Also for consistency, all our source-related tables start with `mz_source`, so we could roll with `mz_source_references`. Do we need an additional qualifier?

Reply:
We do have precedent for a word before `mz_source`, like `mz_kafka_sources`, `mz_postgres_sources`, etc.

Reply:
(I don't mind `mz_source_references` at all though.)